magma/topological_sort.py

```diff
@@ -0,0 +1,241 @@
+        raise Exception("cycle in graph!: node " + curr_node[name_of_node_key])
+    
+    visited_temporary[curr_idx] = True
+
+    for following_step in curr_node[name_of_dependencies_key]:
+        #TODO: can't have duplicates in names with this method!
+        idx_following_node, following_node = find_index_with_given_step_name(graph, name_of_node_key, following_step)
+        topological_sort_DFS_helper(graph, following_node, idx_following_node, name_of_node_key, name_of_dependencies_key, visited_temporary, visited_permanent, queue)
+
+    visited_temporary[curr_idx] = False
+    visited_permanent[curr_idx] = True
+    queue.append(curr_node) 
+    return queue
+    # TODO: for test, can check that there are no duplicates in returned queue   
+
+
+def topological_sort(graph, name_of_node_key, name_of_dependencies_key):
+    """
+    DFS algorithm from wikipedia https://en.wikipedia.org/wiki/Topological_sorting 
+    Logic based on topological sort of directed graph from https://www.geeksforgeeks.org/topological-sorting/
+    TODO: finish this docstring l8r
+    Time complexity = O(V+E), where V = # vertices/nodes (steps), E = # edges (in directed graph, dependencies)
+    https://www.geeksforgeeks.org/detect-cycle-in-directed-graph-using-topological-sort/?id=discuss = cycle detection  :
+    So, in detail, just do a topological sort and get the queue of the results. Then as you pop from the final queue and
+    push to your result vector/array, check all the adjacent nodes of the last popped item and if the adjacent node 
+    exists in the vector then it's a cycle (if A goes to B then B should not precede A in the topological ordering). 
+    ASSUMPTOPN: no self loops (i deletd them)
+
+    pseudocode from wikipedia:
+    L ← Empty list that will contain the sorted nodes
+    while exists nodes without a permanent mark do
+        select an unmarked node n
+        visit(n)
+
+    function visit(node n)
+        if n has a permanent mark then
+            return
+        if n has a temporary mark then
+            stop   (graph has at least one cycle)
+
+        mark n with a temporary mark
+
+        for each node m with an edge from n to m do
+            visit(m)
+
+        remove temporary mark from n
+        mark n with a permanent mark
+        add n to head of L
+    """
+    num_steps = len(graph)
+    visited_temporary = [False]*num_steps
+    visited_permanent = [False]*num_steps
+    queue = [] #First In First Out
+
+    while not all((element == True) for element in visited_permanent):
+        curr_idx = visited_permanent.index(False)
+        curr_node = graph[curr_idx]
+        #calling recursive helper function
+        queue = topological_sort_DFS_helper(graph, curr_node, curr_idx, name_of_node_key, name_of_dependencies_key, visited_temporary, visited_permanent, queue)
+
+    #TODO: remove the following, just for testing
+    for element in queue:
+        print(element["name"])
+
+    return queue
```

Comment:
At least these two functions, and possibly the entire module here, would probably be better suited as a class rather than a collection of functions. There are many variables being passed around the functions that can probably be defined once upon making the class instance and then simply referred to.